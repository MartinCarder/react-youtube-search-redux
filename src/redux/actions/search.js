import { youtubeSearch } from '../services/youtube'
export const SEARCH_REQESTED = 'SEARCH_REQESTED'
export const SEARCH_FAILED = 'SEARCH_FAILED'
export const SEARCH_SUCCESS = 'SEARCH_SUCCESS'

export const searchRequested = searchQuery => {
  return dispatch => {
    return youtubeSearch(searchQuery)
      .then(data => dispatch(searchSuccess(data)))
      .catch(err => {
        dispatch(searchFailed(err))
      })
  }
}

export const searchFailed = message => ({
  type: SEARCH_FAILED,
  message,
})

export const searchSuccess = data => ({
  type: SEARCH_SUCCESS,
  data,
})
