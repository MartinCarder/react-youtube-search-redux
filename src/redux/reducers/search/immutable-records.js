import { Record } from 'immutable'

export const SearchResultRecord = Record(
  {
    id: null,
    title: '',
    publishedAt: '',
    description: '',
    thumbnail: '',
  },
  'SearchResultRecord'
)

export const reviveSearchResultRecord = video =>
  new SearchResultRecord({
    id: video.id.videoId,
    thumbnail: video.snippet.thumbnails.high.url,
    ...video.snippet,
  })
