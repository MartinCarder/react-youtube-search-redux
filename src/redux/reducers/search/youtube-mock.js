export const youtubeMock = [
  {
    id: { kind: 'youtube#video', videoId: 'hY7m5jjJ9mM' },
    snippet: {
      channelId: 'UC9obdDRxQkmn_4YpcBMTYLw',
      channelTitle: 'Tiger FunnyWorks',
      description:
        'Cats are amazing creatures because they make us laugh all the time! Watching funny cats is the hardest try not to laugh challenge! Just look how all these cats & kittens play, fail, get along...',
      liveBroadcastContent: 'none',
      publishedAt: '2017-05-31T09:30:02.000Z',
      thumbnails: {
        high: {
          url: 'https://i.ytimg.com/vi/hY7m5jjJ9mM/hqdefault.jpg',
        },
      },
    },
  },
  {
    id: { kind: 'youtube#video', videoId: 'WEkSYw3o5is' },
    snippet: {
      channelId: 'UC9obdDRxQkmn_4YpcBMTYLw',
      channelTitle: 'Tiger FunnyWorks',
      description:
        'Well well well, cats for you again. But this time, even better, even funnier, even more hilarious! We know you like these furries the most, so thats why we keep making such compilations :)...',
      liveBroadcastContent: 'none',
      publishedAt: '2017-10-07T11:00:02.000Z',
      thumbnails: {
        high: {
          url: 'https://i.ytimg.com/vi/WEkSYw3o5is/hqdefault.jpg',
        },
      },
    },
  },
]
