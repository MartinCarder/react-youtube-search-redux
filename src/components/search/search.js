import React, { PureComponent } from 'react'
import { connect } from 'react-redux'
import { searchRequested } from '../../redux/actions/search'
import SearchBar from './search-bar/search-bar'

export class Search extends PureComponent {
  render() {
    const { startSearch } = this.props
    return (
      <div>
        <SearchBar startSearch={startSearch} />
      </div>
    )
  }
}

const mapDispatchToProps = dispatch => ({
  startSearch: searchTerm => dispatch(searchRequested(searchTerm)),
})
export default connect(null, mapDispatchToProps)(Search)
