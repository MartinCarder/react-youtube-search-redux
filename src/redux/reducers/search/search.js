import { Map, List } from 'immutable'
import { reviveSearchResultRecord } from './immutable-records'
import { SEARCH_REQESTED, SEARCH_FAILED, SEARCH_SUCCESS } from '../../actions/search'

const initialState = Map({
  searchTerm: '',
  status: null,
  data: List(),
  nextPageToken: null,
})

const searchReducer = (state = initialState, action) => {
  switch (action.type) {
    case SEARCH_REQESTED:
      return state.set('status', 'loading')
    case SEARCH_FAILED:
      return state.set('status', 'error')
    case SEARCH_SUCCESS:
      const { data } = action
      return state.withMutations(searchState => {
        searchState.set('data', List(data.items.map(video => reviveSearchResultRecord(video))))
        searchState.set('nextPageToken', data.nextPageToken)
        searchState.set('status', null)
      })
    default:
      return state
  }
}

export default searchReducer
