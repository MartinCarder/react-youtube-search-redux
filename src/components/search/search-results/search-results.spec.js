import React from 'react'
import { shallow } from 'enzyme'
import { List } from 'immutable'
import SearchResults from './search-results'
import { SearchResultRecord } from '../../../redux/reducers/search/immutable-records'

describe('<SearchResults>', () => {
  const results = List([
    SearchResultRecord({
      id: '1234',
      title: 'Another cat video',
      description: 'Cats cats cats!!',
      publishedAt: '2018-03-24T09:30:02.000Z',
    }),
    SearchResultRecord({
      id: '1235',
      title: 'Another cat video Two',
      description: 'More cats cats!!',
      publishedAt: '2018-03-25T09:30:02.000Z',
    }),
  ])

  it('component renders results', () => {
    const wrapper = shallow(<SearchResults selectVideo={jest.fn()} results={results} />)
    expect(wrapper).toMatchSnapshot()
  })

  it('component styles', () => {
    const wrapper = shallow(<SearchResults selectVideo={jest.fn()} results={results} />).dive()
    expect(wrapper).toMatchSnapshot()
  })
})
