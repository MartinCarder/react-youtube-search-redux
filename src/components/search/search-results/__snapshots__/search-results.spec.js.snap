// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchResults> component renders results 1`] = `
<styled.div>
  <VideoItem
    key="1234"
    selectVideo={[Function]}
    video={
      Object {
        "description": "Cats cats cats!!",
        "id": "1234",
        "publishedAt": "2018-03-24T09:30:02.000Z",
        "thumbnail": "",
        "title": "Another cat video",
      }
    }
  />
  <VideoItem
    key="1235"
    selectVideo={[Function]}
    video={
      Object {
        "description": "More cats cats!!",
        "id": "1235",
        "publishedAt": "2018-03-25T09:30:02.000Z",
        "thumbnail": "",
        "title": "Another cat video Two",
      }
    }
  />
</styled.div>
`;

exports[`<SearchResults> component styles 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: row;
  -ms-flex-flow: row;
  flex-flow: row;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-align-content: flex-start;
  -ms-flex-line-pack: start;
  align-content: flex-start;
  margin-top: 51px;
}

<div
  className="c0"
>
  <VideoItem
    key="1234"
    selectVideo={[Function]}
    video={
      Object {
        "description": "Cats cats cats!!",
        "id": "1234",
        "publishedAt": "2018-03-24T09:30:02.000Z",
        "thumbnail": "",
        "title": "Another cat video",
      }
    }
  />
  <VideoItem
    key="1235"
    selectVideo={[Function]}
    video={
      Object {
        "description": "More cats cats!!",
        "id": "1235",
        "publishedAt": "2018-03-25T09:30:02.000Z",
        "thumbnail": "",
        "title": "Another cat video Two",
      }
    }
  />
</div>
`;
